test/dialect/postgresql/test_dialect.py

```diff
@@ -240,53 +241,156 @@ def test_psycopg2_nonempty_connection_string_w_query(self):
         ),
         (
             "postgresql+psycopg2://USER:PASS@/DB"
-            "?host=hostA&host=hostB:portB&host=hostC:portC",
+            "?host=hostA&host=hostB:222&host=hostC:333",
             {
                 "dbname": "DB",
                 "user": "USER",
                 "password": "PASS",
                 "host": "hostA,hostB,hostC",
-                "port": ",portB,portC",
+                "port": ",222,333",
             },
         ),
         (
             "postgresql+psycopg2://USER:PASS@/DB?"
-            "host=hostA:portA&host=hostB:portB&host=hostC:portC",
+            "host=hostA:111&host=hostB:222&host=hostC:333",
             {
                 "dbname": "DB",
                 "user": "USER",
                 "password": "PASS",
                 "host": "hostA,hostB,hostC",
-                "port": "portA,portB,portC",
+                "port": "111,222,333",
             },
         ),
         (
             "postgresql+psycopg2:///"
-            "?host=hostA:portA&host=hostB:portB&host=hostC:portC",
-            {"host": "hostA,hostB,hostC", "port": "portA,portB,portC"},
+            "?host=hostA:111&host=hostB:222&host=hostC:333",
+            {"host": "hostA,hostB,hostC", "port": "111,222,333"},
         ),
         (
             "postgresql+psycopg2:///"
-            "?host=hostA:portA&host=hostB:portB&host=hostC:portC",
-            {"host": "hostA,hostB,hostC", "port": "portA,portB,portC"},
+            "?host=hostA:111&host=hostB:222&host=hostC:333",
+            {"host": "hostA,hostB,hostC", "port": "111,222,333"},
         ),
         (
             "postgresql+psycopg2:///"
-            "?host=hostA,hostB,hostC&port=portA,portB,portC",
-            {"host": "hostA,hostB,hostC", "port": "portA,portB,portC"},
+            "?host=hostA,hostB,hostC&port=111,222,333",
+            {"host": "hostA,hostB,hostC", "port": "111,222,333"},
         ),
-        argnames="url_string,expected",
+        argnames="url_string,expected_psycopg",
     )
     @testing.combinations(
         psycopg2_dialect.dialect(),
         psycopg_dialect.dialect(),
         argnames="dialect",
     )
-    def test_psycopg_multi_hosts(self, dialect, url_string, expected):
+    def test_multi_hosts(self, dialect, url_string, expected_psycopg):
+        url_string = url_string.replace("psycopg2", dialect.name)
+
+        u = url.make_url(url_string)
+
+        if dialect.driver in ("psycopg", "psycopg2"):
+            cargs, cparams = dialect.create_connect_args(u)
+            eq_(cparams, expected_psycopg)
+            eq_(cargs, [])
+        else:
+            assert False
+
+    @testing.combinations(
+        (
+            "postgresql+asyncpg://USER:PASS@/DB?"
+            "host=hostA:111&host=hostB:222&host=hostC:333",
+            {
+                "database": "DB",
+                "user": "USER",
```

Comment:
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755